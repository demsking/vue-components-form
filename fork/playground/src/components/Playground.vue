<template>
  <div class="playground">
    <div class="playground__input">
      <div class="playground__input__schema">
        <h2>Schema</h2>
        <div class="playground__input__schema__editor">
          <PrismEditor
            :code="rawSchema"
            language="json"
            @change="setRawSchema"/>
        </div>
      </div>
      <div class="playground__input__rendering">
        <h2>Rendering</h2>
        <div ref="form" class="playground__input__rendering__viewport">
          <FormSchema :schema="schema" :descriptor="descriptor" v-model="model" @ready="generateCode" @submit.prevent>
            <div class="buttons">
              <button type="submit">Subscribe</button>
            </div>
          </FormSchema>
        </div>
      </div>
      <div class="playground__input__model">
        <h2>Model</h2>
        <pre>{{ model }}</pre>
      </div>
    </div>
    <div class="playground__output">
      <h2>Generated HTML</h2>
      <div class="playground__output__code">
        <PrismEditor :code="code" language="html" readonly/>
      </div>
    </div>
  </div>
</template>

<script>
  import '../libs/Prism';
  import '../styles/Prism.css';

  import { html } from 'js-beautify';

  import PrismEditor from 'vue-prism-editor';
  import FormSchema from '../../../dist/FormSchema.esm';
  import Schema from '../schema/newsletter';

  export default {
    components: { PrismEditor, FormSchema },
    data: () => ({
      model: undefined,
      rawSchema: JSON.stringify(Schema, null, 2),
      code: '',
      formatter: null,
      descriptor: {
        properties: {
          name: {
            properties: {
              first_name: {
                attrs: {
                  placeholder: 'Your First Name',
                  title: 'Please enter your first name'
                }
              },
              middle_name: {
                attrs: {
                  placeholder: 'Your Middle Name',
                  title: 'Please enter your middle name'
                }
              },
              last_name: {
                attrs: {
                  placeholder: 'Your Last Name',
                  title: 'Please enter your last name'
                }
              }
            }
          },
          email: {
            attrs: {
              placeholder: 'Your Email',
              title: 'Please enter your email'
            }
          },
          day: {
            label: 'Newsletter Day',
            attrs: {
              placeholder: 'Select your list subscription',
              title: 'Please select your list subscription'
            },
            labels: {
              monday: 'Monday',
              tuesday: 'Tuesday',
              wednesday: 'Wednesday',
              thursday: 'Thursday',
              friday: 'Friday',
              saturday: 'Saturday',
              sunday: 'Sunday'
            }
          },
          source: {
            kind: 'textarea',
            label: 'Source',
            description: 'Ex. Using the NPM Search Engine',
            attrs: {
              placeholder: 'How did you hear about us?'
            }
          },
          password: {
            attrs: {
              type: 'password'
            }
          },
          frequence: {
            labels: {
              daily: 'Daily',
              weekly: 'Weekly'
            }
          }
        }
      }
    }),
    computed: {
      schema() {
        return JSON.parse(this.rawSchema);
      }
    },
    methods: {
      setRawSchema(value) {
        try {
          JSON.parse(value);

          this.rawSchema = value;
        } catch(e) {}
      },
      generateCode() {
        const code = this.$refs.form.innerHTML;

        this.code = html(code.replace(/\s*<!---->/g, ''));
      }
    }
  }
</script>

<style>
  .playground {
    text-align: left;
    margin: auto;
    /* display: flex; */
    flex-direction: column;
    color: #f5f5f5;
    background-color: #333333;
    overflow: hidden;

    position: relative;
    height: 100%;
  }

  .playground h2 {
    margin: 0;
    padding: 10px;
    font-size: .90em;
    font-weight: 400;
    width: 100%;
  }

  .playground pre {
    margin: 0;
  }

  .playground__input {
    display: flex;
    flex-direction: row;
    /* max-height: 550px; */

    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 60%;
  }

  .playground h2,
  .playground__input {
    border-bottom: 1px solid rgba(255, 255, 255, .1);
  }

  .playground__input__schema {
    display: flex;
    flex-direction: column;
  }

  .playground__input__schema__editor {
    flex: 1;
    display: flex;
    align-items: flex-start;
    max-width: 500px;
    overflow: auto;
  }

  .playground__input__rendering {
    border-left: 1px solid rgba(255, 255, 255, .1);
    border-right: 1px solid rgba(255, 255, 255, .1);
  }

  .playground__input__rendering,
  .playground__input__model {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    overflow: hidden;
  }

  .playground__input__rendering__viewport {
    padding: 20px;
    overflow: auto;
    margin: auto;
    color: #333;
    background-color: #f5f5f5;
  }

  .playground__input__rendering legend {
    font-size: 1.7em;
    text-align: center;
    margin-top: 0;
    margin-bottom: .2em
  }

  .playground__input__rendering legend + p {
    display: block;
    text-align: center;
    margin-bottom: 1.2em
  }

  .playground__input__rendering input,
  .playground__input__rendering textarea,
  .playground__input__rendering select {
    display: block;
  }

  .playground__input__model {
    margin: 0;
    min-width: 100px;
  }

  .playground__output {
    position: absolute;
    top: 60%;
    left: 0;
    width: 100%;
    height: 40%;

    overflow: hidden;
    display: flex;
    flex-direction: column;
  }

  .playground__output__code {
    flex: 1;
    display: flex;
    overflow: auto;
  }

  [data-fs-field] {
    display: flex;
    flex-direction: row;
    margin-bottom: 5px;
  }

  [data-fs-field] label {
    display: block;
    width: 120px;
    text-align: right;
    margin-right: 10px
  }

  [data-fs-kind="enum"] [data-fs-field] {
    flex-direction: row-reverse;
  }

  [data-fs-kind="enum"] [data-fs-field] label {
    text-align: left;
    flex: 1;
  }
</style>
