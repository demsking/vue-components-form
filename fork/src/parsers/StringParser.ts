import { AbstractParser } from '@/parsers/AbstractParser';
import { StringField, ScalarDescriptor, FieldKind } from '@/types';

export class StringParser extends AbstractParser<string, ScalarDescriptor, StringField> {
  get kind(): FieldKind {
    return this.isEnum ? 'radio' : 'string';
  }

  get type() {
    if (this.field.attrs.input.type) {
      return this.field.attrs.input.type;
    }

    if (this.isEnum) {
      return 'radio';
    }

    switch (this.schema.format) {
      case 'date':
        return 'date';

      case 'date-time':
        return 'datetime-local';

      case 'email':
      case 'idn-email':
        return 'email';

      case 'time':
        return 'time';

      case 'uri':
        return 'url';

      default:
        return 'text';
    }
  }

  parse() {
    this.parseField();

    this.field.attrs.input.value = this.field.model;
    this.field.attrs.input.minlength = this.schema.minLength;
    this.field.attrs.input.maxlength = this.schema.maxLength;
    this.field.attrs.input.pattern = this.schema.pattern;
  }

  parseValue(data: any): string {
    return typeof data !== 'undefined' ? `${data}` : '';
  }
}
