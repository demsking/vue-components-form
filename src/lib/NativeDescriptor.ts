import { Schema } from '@/lib/Schema';
import { JsonSchema } from '@/types/jsonschema';
import { NativeElements } from '@/lib/NativeElements';

import {
  FieldKind,
  ScalarDescriptor,
  ObjectDescriptor,
  ArrayDescriptor,
  DescriptorInstance
} from '@/types';

export const NativeDescriptor = {
  kind(schema: JsonSchema) {
    return schema.enum
      ? schema.enum.length > 4 ? 'list' : 'enum'
      : schema.type;
  },
  get<T = DescriptorInstance>(schema: JsonSchema, kind?: FieldKind): T {
    const kindUsed = typeof kind === 'undefined'
      ? NativeDescriptor.kind(schema)
      : kind;

    const element = NativeElements.get(kindUsed);

    if (Schema.isScalar(schema)) {
      const descriptor: ScalarDescriptor = {
        kind: kindUsed,
        label: schema.title,
        helper: schema.description,
        component: element,
        attrs: {},
        props: {},
        items: {}
      };

      return descriptor as T;
    }

    if (schema.type === 'array') {
      const descriptor: ArrayDescriptor = {
        kind: kindUsed,
        label: schema.title,
        helper: schema.description,
        component: element,
        attrs: {},
        props: {},
        items: [],
        addButtonLabel: '+'
      };

      return descriptor as unknown as T;
    }

    const descriptor: ObjectDescriptor = {
      kind: kindUsed,
      label: schema.title,
      helper: schema.description,
      component: element,
      attrs: {},
      props: {},
      order: []
    };

    return descriptor as T;
  }
};
