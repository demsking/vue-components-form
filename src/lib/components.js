import { INPUT_TYPES } from './parser'

const BLOCK_TYPES = [INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT]

const Input = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const element = BLOCK_TYPES.includes(field.attrs.type)
      ? h(field.attrs.type, data, slots().default)
      : h('input', data)
    const nodes = [ element ]

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label || (field.isArrayField && !BLOCK_TYPES.includes(field.attrs.type))) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const CheckboxGroup = {
  functional: true,
  render (h, { data, slots }) {
    const name = data.field.attrs.name
    const description = data.field.description

    const children = [ ...slots().default ]

    if (description) {
      children.unshift(h('legend', description))

      delete data.field.description
    }

    return h('fieldset', { attrs: { name } }, children)
  }
}

const ArrayInputs = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-array-inputs': true
      }
    }, slots().default)
  }
}

const Label = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const attrs = {
      'data-fs-field': field.attrs.id,
      'data-fs-required': field.attrs.required
    }

    return h('div', { attrs }, [
      h('label', {
        attrs: {
          for: field.attrs.id
        }
      }, field.label),
      h('div', {
        attrs: {
          'data-fs-field-input': field.attrs.id
        }
      }, slots().default)
    ])
  }
}

const Fieldset = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const newItemButton = data.newItemButton
    const nodes = []

    if (field.isArrayField) {
      if (newItemButton) {
        nodes.push(h(ArrayInputs, slots().default))
        nodes.push(h(ArrayButton, newItemButton))
      } else {
        nodes.push(h(CheckboxGroup, data, slots().default))
      }
    } else {
      nodes.push(h('fieldset', slots().default))
    }

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const ArrayButton = {
  functional: true,
  render (h, { data }) {
    return h('button', {
      attrs: { type: 'button', ...data.props },
      on: data.on
    }, 'Add')
  }
}

const ErrorElement = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-error': true
      }
    }, slots().default)
  }
}

const TAGS = {
  title: 'h1',
  description: 'p',
  formwrapper: 'div',
  form: 'form',
  text: Input,
  option: 'option',
  fieldset: Fieldset,
  error: ErrorElement
}

export class Components {
  constructor () {
    this.$ = {}

    for (let type in TAGS) {
      this.set(type, TAGS[type], true)
    }
  }

  set (type, component, native = false) {
    this.$[type] = { type, native, component }
  }

  input ({ field, fieldParent = null, listeners = {} }) {
    const { type } = field.attrs
    const element = field.hasOwnProperty('items') && groupedArrayTypes.includes(type)
      ? this.$.fieldset
      : this.$[type] || this.$[INPUT_TYPES.TEXT]

    const data = {
      field,
      fieldParent,
      components: this.components,
      attrs: {},
      props: {},
      domProps: {},
      [argName(element)]: { ...field.attrs },
      on: listeners
    }

    return { element, data }
  }
}

export function argName (el) {
  return el.native ? 'attrs' : 'props'
}

export const groupedArrayTypes = [
  INPUT_TYPES.RADIO, INPUT_TYPES.CHECKBOX, 'input', INPUT_TYPES.TEXTAREA
]

export const fieldTypesAsNotArray = [
  INPUT_TYPES.RADIO, INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT
]

export const inputName = (field, index) => `${field.attrs.name}-${index}`
