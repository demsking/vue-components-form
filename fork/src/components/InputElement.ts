import { CreateInput } from '@/lib/NativeElementsLib';
import { FieldElement } from "@/components/FieldElement";
import { InputComponent, InputField } from '@/types';

export const InputElement: InputComponent = {
  name: 'InputElement',
  functional: true,
  render(h, { data }) {
    return h(FieldElement, data, [
      CreateInput<InputField>(h, 'input', data)
    ]);
  }
};
