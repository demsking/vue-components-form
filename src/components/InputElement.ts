import { CreateInput } from '@/lib/CreateInput';
import { FieldElement } from '@/components/FieldElement';
import { InputComponent, InputField } from '@/types';

export const InputElement: InputComponent = {
  name: 'InputElement',
  functional: true,
  render(h, { data, children }) {
    return h(FieldElement, data, [
      CreateInput<InputField>(h, 'input', data),
      ...children
    ]);
  }
};
