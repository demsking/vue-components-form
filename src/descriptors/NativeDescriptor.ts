import { Schema } from '@/lib/Schema';
import { JsonSchema } from '@/types/jsonschema';
import { NativeElements } from '@/lib/NativeElements';

import {
  FieldKind,
  ScalarDescriptor,
  ObjectDescriptor,
  ArrayDescriptor,
  DescriptorInstance
} from '@/types';

export const NativeDescriptor = {
  get<T = DescriptorInstance>(schema: JsonSchema, kind?: FieldKind): T {
    const element = NativeElements.get(kind || schema.type);

    if (Schema.isScalar(schema)) {
      const descriptor: ScalarDescriptor = {
        label: schema.title,
        description: schema.description,
        component: element,
        attrs: {},
        props: {},
        labels: {}
      };

      return descriptor as T;
    }

    if (schema.type === 'array') {
      const descriptor: ArrayDescriptor = {
        label: schema.title,
        description: schema.description,
        component: element,
        attrs: {},
        props: {},
        items: [],
        addButtonLabel: '+'
      };

      return descriptor as unknown as T;
    }

    const descriptor: ObjectDescriptor = {
      label: schema.title,
      description: schema.description,
      component: element,
      attrs: {},
      props: {},
      order: []
    };

    return descriptor as T;
  }
};
