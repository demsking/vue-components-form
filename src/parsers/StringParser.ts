import { Parser } from '@/parsers/Parser';
import { FieldKind, Dictionary, ScalarDescriptor, StringAttributes } from '@/types';
import { Pattern } from '@/lib/Pattern';

const TypeFormat: Dictionary<string> = {
  date: 'date',
  'date-time': 'datetime-local',
  email: 'email',
  'idn-email': 'email',
  time: 'time',
  uri: 'url'
};

export class StringParser extends Parser<'string', string, StringAttributes, ScalarDescriptor> {
  get kind(): FieldKind {
    return this.isEnumItem ? 'radio' : 'string';
  }

  get type(): string {
    if (this.isEnumItem) {
      return 'radio';
    }

    return this.schema.format
      ? TypeFormat[this.schema.format]
      : 'text';
  }

  parse() {
    super.parse();

    this.attrs.value = this.field.input.value;
    this.attrs.minlength = this.schema.minLength;
    this.attrs.maxlength = this.schema.maxLength;

    if (this.schema.pattern) {
      this.attrs.pattern = this.schema.pattern;
    } else if (this.schema.hasOwnProperty('const')) {
      this.attrs.pattern = Pattern.escape(`${this.schema.const}`);
    }

    this.commit();
  }

  parseValue(data: unknown): string | undefined {
    return typeof data !== 'undefined' ? `${data}` : undefined;
  }
}

Parser.register('string', StringParser);
