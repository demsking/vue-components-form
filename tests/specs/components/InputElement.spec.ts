import { mount } from '@vue/test-utils';
import { InputElement } from '@/components/InputElement';
import { StringParser } from '@/parsers/StringParser';
import { Dictionary, ScalarDescriptor, ParserOptions } from '@/types';
import { NativeDescriptor } from '@/lib/NativeDescriptor';

const options: ParserOptions<string, ScalarDescriptor> = {
  schema: {
    type: 'string',
    pattern: 'arya|jon',
    minLength: 5,
    maxLength: 15,
    title: 'Name',
    description: 'Your full name'
  },
  model: 'Goku',
  id: 'id-name',
  name: 'name',
  descriptorConstructor: NativeDescriptor.get
};

const parser = new StringParser(options);

parser.parse();

describe('components/InputElement', () => {
  it('should successfully render component', () => {
    const context: any = {
      attrs: parser.field.attrs.input,
      props: {
        field: parser.field
      }
    };

    const wrapper = mount(InputElement, { context });
    const expected = '<div data-fs-kind="string" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label><div data-fs-input="text"><input id="id-name" type="text" name="name" aria-labelledby="id-name-label" aria-describedby="id-name-desc" value="Goku" minlength="5" maxlength="15" pattern="arya|jon"><p id="id-name-desc">Your full name</p></div></div>';

    expect(wrapper.html()).toBe(expected);
  });
});
