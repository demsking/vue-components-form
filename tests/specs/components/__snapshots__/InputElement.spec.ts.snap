// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InputElement should successfully render component: <div data-fs-kind="string" data-fs-type="text" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-name" type="text" name="name" value="Goku" pattern="arya|jon" minlength="5" maxlength="15" aria-labelledby="id-name-label" aria-describedby="id-name-helper"></div><p id="id-name-helper">Your full name</p></div></div> 1`] = `
<div data-fs-kind="string" data-fs-type="text" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label>
  <div data-fs-wrapper="2">
    <div data-fs-input="text"><input id="id-name" type="text" name="name" aria-labelledby="id-name-label" aria-describedby="id-name-helper" value="Goku" pattern="arya|jon" minlength="5" maxlength="15"></div>
    <p id="id-name-helper">Your full name</p>
  </div>
</div>
`;
